### PR TITLE
[4.0] Allow jForm to support data attribute

### PR DESCRIPTION
**Pull request by vaivk369**
#16141

### CODE REVIEW

File: libraries/src/Form/FormField.php [hunk 1]
```diff
@@ -546,13 +563,21 @@ public function __set($name, $value)
...
 				{
-					Log::add("Cannot access protected / private property $name of " . __CLASS__);
+					$this->dataAttributes[$name] = $value;
 				}
 				else
 				{
-					$this->$name = $value;
+					if (property_exists(__CLASS__, $name))
+					{
+						\JLog::add("Cannot access protected / private property $name of " . __CLASS__);
```
Comment (alikon) [inline libraries/src/Form/FormField.php]:
```suggestion
						Log::add("Cannot access protected / private property $name of " . __CLASS__);
```

**Review by alikon — CHANGES_REQUESTED** [review]

**Comment by alikon** [comment]
is this a work in progress ?
as example this one has not been added  from #16141
https://github.com/joomla/joomla-cms/blob/4.0-dev/layouts/joomla/form/field/color/advanced.php

File: layouts/joomla/form/field/calendar.php [hunk 2]
```diff
@@ -80,6 +81,13 @@
 	$attributes['required'] = '';
 }
 
+// Data attributes - data-*
+$dataAttribute = '';
+foreach ($dataAttributes as $key => $attrValue)
```
Comment (Quy) [inline layouts/joomla/form/field/calendar.php]:
```suggestion

foreach ($dataAttributes as $key => $attrValue)
```

Repeat with other instances.

**Comment by dgrammatiko** [comment]
@twsvaishali please extend this for `aria-*`

**Comment by vaivk369** [comment]
> is this a work in progress ?
> as example this one has not been added from #16141
> https://github.com/joomla/joomla-cms/blob/4.0-dev/layouts/joomla/form/field/color/advanced.php

Have added remaining fields in latest commits. Please check.

**Comment by vaivk369** [comment]
> @twsvaishali please extend this for `aria-*`

Can we start new PR for this?

**Comment by alikon** [comment]
> Have added remaining fields in latest commits. Please check.

seems fine now thanks

yes maybe another PR  for the `aria-*`

**Comment by dgrammatiko** [comment]
@alikon @twsvaishali for the `aria-*` all is needed is a var ariaAttributes
and adding 
```php
				// Check for aria attribute
 				if (strpos($name, "aria-") === 0 && array_key_exists($name, $this->ariaAttributes))
 				{
 					return $this->ariaAttributes[$name];
 				}
```
in the main class and some minor adjustments in the layouts. Of course it can be done in another PR to make things easier for testing. I just had to mention that, since a11y is always an afterthought...

**Comment by dgrammatiko** [comment]
@twsvaishali since you're redoing this maybe we could change the layouts logic to:
```php
if (!empty($dataAttributes)) {
	// Data attributes - data-*
	$dataAttribute = '';

	foreach ($dataAttributes as $key => $attrValue)
	{
		$dataAttribute .= ' ' . $key . '="' . htmlspecialchars($attrValue, ENT_COMPAT, 'UTF-8') . '"';
	}
}
```

In essence the loop will execute **only** if there are any `data-*` attributes defined.
Micro optimisation? Maybe, maybe not, I haven't benchmarked the code but this seems reasonable...

**Comment by vaivk369** [comment]
@dgrammatiko Yes. Will start another PR  for the aria-* once this is tested and merged. Also I am actually not sure if all fields will have area attribute.

File: layouts/joomla/form/field/calendar.php [hunk 3]
```diff
@@ -80,6 +81,14 @@
 	$attributes['required'] = '';
 }
 
+// Data attributes - data-*
```
Comment (Quy) [inline layouts/joomla/form/field/calendar.php]: I would remove the comment as it is mentioned in the doc block.

File: layouts/joomla/form/field/password.php [hunk 4]
```diff
@@ -83,6 +92,7 @@
 	!empty($minUppercase) ? 'data-min-uppercase="' . $minUppercase . '"' : '',
 	!empty($minLowercase) ? 'data-min-lowercase="' . $minLowercase . '"' : '',
 	!empty($forcePassword) ? 'data-min-force="' . $forcePassword . '"' : '',
+	$dataAttribute
```
Comment (Quy) [inline layouts/joomla/form/field/password.php]:
```suggestion
	$dataAttribute,
```

File: layouts/joomla/form/field/text.php [hunk 5]
```diff
@@ -53,12 +54,19 @@
 	$list = 'list="' . $id . '_datalist"';
 }
 
+$dataAttribute = '';
+foreach ($dataAttributes as $key => $attrValue)
```
Comment (Quy) [inline layouts/joomla/form/field/text.php]:
```suggestion

foreach ($dataAttributes as $key => $attrValue)
```

File: layouts/joomla/form/field/time.php [hunk 6]
```diff
@@ -57,7 +66,8 @@
 	isset($step) ? 'step="' . $step . '"' : '',
 	isset($min) ? 'min="' . $min . '"' : '',
 	$required ? 'required' : '',
-	$autofocus ? 'autofocus' : ''
+	$autofocus ? 'autofocus' : '',
+	$dataAttribute
```
Comment (Quy) [inline layouts/joomla/form/field/time.php]:
```suggestion
	$dataAttribute,
```

File: layouts/joomla/form/field/moduleorder.php [hunk 7]
```diff
@@ -76,6 +77,14 @@
 	$attributes['onchange'] = 'onchange="' . $onchange . '"';
 }
 
+// Data attributes - data-*
+$dataAttribute = '';
```
Comment (Quy) [inline layouts/joomla/form/field/moduleorder.php]: Variable not used.

File: layouts/joomla/form/field/calendar.php [hunk 8]
```diff
@@ -48,6 +48,7 @@
  * @var   array    $checkedOptions  Options that will be set as checked.
  * @var   boolean  $hasValue        Has this field a value assigned?
  * @var   array    $options         Options available for this field.
+ * @var   array    $dataAttribute   Miscellaneous data attribute for eg, data-*.
```
Comment (SharkyKZ) [inline layouts/joomla/form/field/calendar.php]: This should be variable from extracted data. So `$dataAttributes` not `$dataAttribute`. Same in other fields.

Comment (vaivk369) [reply layouts/joomla/form/field/calendar.php]: done. Changes committed.

File: layouts/joomla/form/field/moduleorder.php (hunk 7 above)
Comment (vaivk369) [reply layouts/joomla/form/field/moduleorder.php]: done. Changes committed.

File: layouts/joomla/form/field/time.php (hunk 6 above)
Comment (vaivk369) [reply layouts/joomla/form/field/time.php]: done. Changes committed.

File: libraries/src/Form/FormField.php [hunk 9]
```diff
@@ -622,6 +647,16 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
 			$this->value = $value;
 		}
 
+		// Lets detect miscellaneous data attribute. For eg, data-*
+		foreach ($this->element->attributes() as $key => $value)
+		{
+			if (strpos($key, "data-") === 0)
```
Comment (Quy) [inline libraries/src/Form/FormField.php]:
```suggestion
			if (strpos($key, 'data-') === 0)
```

File: libraries/src/Form/FormField.php [hunk 10]
```diff
@@ -546,13 +563,21 @@ public function __set($name, $value)
 				break;
 
 			default:
-				if (property_exists(__CLASS__, $name))
+				// Detect data attribute(s)
+				if (strpos($name, "data-") === 0)
```
Comment (Quy) [inline libraries/src/Form/FormField.php]:
```suggestion
				if (strpos($name, 'data-') === 0)
```

File: layouts/joomla/form/field/combo.php [hunk 11]
```diff
@@ -70,4 +81,5 @@
 	value="<?php echo htmlspecialchars($value, ENT_COMPAT, 'UTF-8'); ?>"
 	<?php echo $attr; ?>
 	data-list="<?php echo implode(', ', $val); ?>"
+	"<?php echo $dataAttribute; ?>"
```
Comment (Quy) [inline layouts/joomla/form/field/combo.php]:
```suggestion
	<?php echo $dataAttribute; ?>
```

File: libraries/src/Form/FormField.php [hunk 12]
```diff
@@ -453,6 +463,13 @@ public function __get($name)
 
 			case 'title':
 				return $this->getTitle();
+
+			default:
+				// Check for data attribute
+				if (strpos($name, "data-") === 0 && array_key_exists($name, $this->dataAttributes))
```
Comment (Quy) [inline libraries/src/Form/FormField.php]:
```suggestion
				if (strpos($name, 'data-') === 0 && array_key_exists($name, $this->dataAttributes))
```

File: layouts/joomla/form/field/email.php [hunk 13]
```diff
@@ -45,8 +45,19 @@
  * @var   array    $inputType       Options available for this field.
  * @var   array    $spellcheck      Options available for this field.
  * @var   string   $accept          File types that are accepted.
+ * @var   array    $dataAttributes   Miscellaneous data attribute for eg, data-*.
```
Comment (Quy) [inline layouts/joomla/form/field/email.php]:
```suggestion
 * @var   array    $dataAttributes  Miscellaneous data attribute for eg, data-*.
```

**Comment by wilsonge** [comment]
Rebased this. Anyone able to test this?

File: layouts/joomla/form/field/user.php [hunk 14]
```diff
@@ -99,6 +100,14 @@
 {
 	$inputAttributes['placeholder'] = Text::_('JLIB_FORM_SELECT_USER');
 }
+
+if (!empty($dataAttributes))
+{
+	foreach ($dataAttributes as $key => $attrValue)
+	{
+		$inputAttributes[$key] .= htmlspecialchars($attrValue, ENT_COMPAT, 'UTF-8');
```
Comment (Quy) [inline layouts/joomla/form/field/user.php]:
```suggestion
		$inputAttributes[$key] = htmlspecialchars($attrValue, ENT_COMPAT, 'UTF-8');
```

Comment (richard67) [reply layouts/joomla/form/field/user.php]: @wilsonge Shall we just commit this suggestion by @Quy ? Or ask the author of this PR to do it?

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on d46baf19504b0463f348218b9a5656dd73c61891<br /><br />Works so far.

Maybe some fields are missing because they did not exist (yet) when this PR was made.

It needs someone who checks for completeness and if necessary makes a new PR.

Same with the idea to do the same for "aria-" attributes, this should be done in another PR, too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27212">issues.joomla.org/tracker/joomla-cms/27212</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on d46baf19504b0463f348218b9a5656dd73c61891<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27212">issues.joomla.org/tracker/joomla-cms/27212</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27212">issues.joomla.org/tracker/joomla-cms/27212</a>.</sub>

**Comment by richard67** [comment]
@wilsonge @Quy What about the following field layouts? They are not touched by this PR. Shall they be done, too?
- checkboxes.php
- color/slider.php
- contenthistory.php
- radio/buttons.php
- radio/switcher.php
- tag.php

**Comment by HLeithner** [comment]
Thanks all!

**Pull request merged**